lib/pool.dart

```diff
@@ -121,6 +121,44 @@ class Pool {
     });
   }
 
+  /// Requests a resource for the combined duration of the [callback] and the
+  /// processing of the Stream it returns.
+  /// (E.g. a database connection is hold until the the results get processed.)
+  Stream/*<T>*/ streamWithResource/*<T>*/(Stream/*<T>*/ callback()) {
+    StreamController/*<T>*/ streamController;
+    StreamSubscription/*<T>*/ streamSubscription;
+    Completer resourceCompleter = new Completer();
+    final complete = () {
+      if (streamSubscription != null) {
+        streamSubscription.cancel();
+        streamSubscription = null;
+      }
+      if (!resourceCompleter.isCompleted) {
+        resourceCompleter.complete();
+        streamController.close();
+      }
+    };
+    streamController = new StreamController/*<T>*/(onCancel: complete);
+    withResource(() {
+      if (resourceCompleter.isCompleted) return null;
+      new Future.sync(callback).then((Stream/*<T>*/ stream) {
+        if (stream == null) {
+          complete();
+          return null;
+        }
+        streamSubscription = stream.listen(streamController.add,
+            onError: streamController.addError,
+            onDone: complete,
+            cancelOnError: true);
```

Comment:
Rather than manually forwarding events, use [`new StreamTransformer.fromHandlers()`](https://api.dartlang.org/stable/1.21.0/dart-async/StreamTransformer/StreamTransformer.fromHandlers.html) to release the resource on a "done" event. That'll ensure that the resulting stream is *cancel-correct*, meaning that it properly forwards cancellations and pauses to the underlying stream.